### PR TITLE
[FEAT]:Implement backend support for region selection #12

### PR DESCRIPTION
**Pull request by unyimeudemy**
# Implement Region Management Endpoints

## Summary
This pull request introduces the `RegionController` for managing regions in the application. It includes endpoints for retrieving all regions, assigning a region to a user, retrieving a user's assigned region, and updating a user's region. The controller leverages the `RegionService` for business logic and uses DTOs for data transfer.

### Endpoints Implemented

- **`GET /api/v1/regions`**: Retrieves a list of all available regions.
  - **Success Response (200 OK)**: Returns a list of `RegionDto` objects.
  - **Error Responses**:
    - **404 Not Found**: If no regions are available.
    - **400 Bad Request**: For other errors during retrieval.

- **`POST /api/v1/regions`**: Assigns a region to a user based on the provided `UserRegionDto`.
  - **Success Response (200 OK)**: Returns `UserRegionResponseDto` with assigned region details.
  - **Error Responses**:
    - **404 Not Found**: If the specified region is not available.
    - **401 Unauthorized**: For unauthorized access.
    - **409 Conflict**: If the region is already assigned to the user.
    - **400 Bad Request**: For other errors during assignment.

- **`GET /api/v1/regions/{user_id}`**: Retrieves the region assigned to a specific user.
  - **Success Response (302 Found)**: Returns `UserRegionDto` with the assigned region details.
  - **Error Responses**:
    - **404 Not Found**: If no region is assigned to the user.
    - **400 Bad Request**: For invalid parameters.

- **`PUT /api/v1/regions/{user_id}`**: Updates the region assigned to a specific user.
  - **Success Response (200 OK)**: Returns `UpdateUserRegionDto` with updated region details.
  - **Error Responses**:
    - **400 Bad Request**: For invalid region choices or other errors.
    - **401 Unauthorized**: For unauthorized updates.
    - **404 Not Found**: If no region is assigned to the user.

### OpenAPI Annotations
- Added `@Operation` and `@ApiResponses` annotations to document each endpoint with Swagger.
- Included details for successful operations and potential error responses.

### Error Handling
- Configured responses for different error scenarios, including not found, bad request, conflict, and unauthorized errors.

## Checklist
- [x] Implemented endpoints for region management.
- [x] Added OpenAPI documentation for each endpoint.
- [x] Configured error handling and response codes.
- [x] Included detailed success and error responses in the API documentation.

## Reference Issue
- [[FEAT]: Implement Backend Support for Region Selection #12](https://github.com/hngprojects/hng_boilerplate_java_web/issues/12)

### CODE REVIEW

**Review by billmal071 — APPROVED** [review]
LGTM

File: src/test/resources/application.properties [hunk 1]
```diff
@@ -0,0 +1,6 @@
+spring.datasource.url=jdbc:h2:mem:testdb;MODE=PostgreSQL;DATABASE_TO_LOWER=TRUE;DEFAULT_NULL_ORDERING=HIGH
```
Comment (NdubuisiJr) [inline src/test/resources/application.properties:1]: Do not check this in to github.

File: src/main/java/hng_java_boilerplate/mappers/Mapper.java [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package hng_java_boilerplate.mappers;
+
+
+public interface Mapper <A,B>{
```
Comment (NdubuisiJr) [inline src/main/java/hng_java_boilerplate/mappers/Mapper.java:4]: Use better name than `A` and `B`

File: src/main/java/hng_java_boilerplate/region/controller/RegionController.java [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+
+
+    @Operation(summary = "Get all regions", description = "Retrieve all available regions")
+    @ApiResponses(value = {
+            @ApiResponse(responseCode = "200", description = "Successful operation, returns a list of regions"),
+            @ApiResponse(responseCode = "404", description = "Not found, no regions available")
+    })
+    @GetMapping()
+    public ResponseEntity<?> getAllRegion(){
+
```
Comment (NdubuisiJr) [inline src/main/java/hng_java_boilerplate/region/controller/RegionController.java:48]: Remove extra space

File: src/main/java/hng_java_boilerplate/region/controller/RegionController.java [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+        }
+
+        return new ResponseEntity<>(
+                RegionErrorResponseDto.builder()
+                        .message("Bad request")
+                        .statusCode("400")
+                        .build(),
+                HttpStatus.BAD_REQUEST
+        );
+
```
Comment (NdubuisiJr) [inline src/main/java/hng_java_boilerplate/region/controller/RegionController.java:81]: Remove extra space

File: src/main/java/hng_java_boilerplate/region/controller/RegionController.java [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+    private final RegionService regionService;
+
+    private final Mapper<UserRegionEntity, UserRegionDto> userRegionMapper;
+
+    @Autowired
+    public RegionController(RegionService regionService, Mapper<UserRegionEntity, UserRegionDto> userRegionMapper) {
+        this.regionService = regionService;
+        this.userRegionMapper = userRegionMapper;
+    }
+
```
Comment (NdubuisiJr) [inline src/main/java/hng_java_boilerplate/region/controller/RegionController.java:39]: Remove extra space

File: src/main/java/hng_java_boilerplate/region/controller/RegionController.java [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+import io.swagger.v3.oas.annotations.media.Schema;
+import io.swagger.v3.oas.annotations.responses.ApiResponse;
+import io.swagger.v3.oas.annotations.responses.ApiResponses;
+import io.swagger.v3.oas.annotations.tags.Tag;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+import org.springframework.security.core.Authentication;
+
```
Comment (NdubuisiJr) [inline src/main/java/hng_java_boilerplate/region/controller/RegionController.java:19]: Remove extra space

File: src/main/java/hng_java_boilerplate/region/controller/RegionController.java [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+                    Optional.of(RegionErrorResponseDto.builder()
+                            .statusCode("404")
+                            .message("Not found")
+                            .build()),
+                    HttpStatus.BAD_REQUEST
+            );
+        }
+
+
+        //========== Ensure that assignee is owner of account. =================
```
Comment (NdubuisiJr) [inline src/main/java/hng_java_boilerplate/region/controller/RegionController.java:112]: Remove extra space and in-line comment.

File: src/main/java/hng_java_boilerplate/region/controller/RegionController.java [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+            return new ResponseEntity<>(
+                    Optional.of(RegionErrorResponseDto.builder()
+                            .statusCode("401")
+                            .message("Unauthorized request")
+                            .build()),
+                    HttpStatus.UNAUTHORIZED
+            );
+        }
+
+
```
Comment (NdubuisiJr) [inline src/main/java/hng_java_boilerplate/region/controller/RegionController.java:125]: Remove extra space

File: src/main/java/hng_java_boilerplate/region/controller/RegionController.java [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+            if(regionService.hasUserAssignedRegion(userRegionDTO.getUserId())){
+                return new ResponseEntity<>(
+                        Optional.of(RegionErrorResponseDto.builder()
+                                .statusCode("409")
+                                .message("Conflict")
+                                .build()),
+                        HttpStatus.CONFLICT
+                );
+            }
+
```
Comment (NdubuisiJr) [inline src/main/java/hng_java_boilerplate/region/controller/RegionController.java:137]: Remove extra space

File: src/main/java/hng_java_boilerplate/region/controller/RegionController.java [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+            return new ResponseEntity<>(
+                    Optional.of(RegionErrorResponseDto.builder()
+                            .statusCode("400")
+                            .message("Bad request")
+                            .build()),
+                    HttpStatus.BAD_REQUEST
+            );
+        }
+    }
+
```
Comment (NdubuisiJr) [inline src/main/java/hng_java_boilerplate/region/controller/RegionController.java:161]: Remove extra space

File: src/main/java/hng_java_boilerplate/region/controller/RegionController.java [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+    @PutMapping("/{user_id}")
+    public ResponseEntity<?> updateUserRegion(
+            @PathVariable("user_id") String userID,
+            @RequestBody RegionUpdateDto regionUpdateDto,
+            Authentication authentication
+    ) {
+
+
+
+//========== Ensure that only the current authenticated users can update their region preferences. ===============================================
```
Comment (NdubuisiJr) [inline src/main/java/hng_java_boilerplate/region/controller/RegionController.java:215]: Remove extra space and inline comments

File: src/main/java/hng_java_boilerplate/region/dto/RegionDto.java [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+
+    @Schema(description = "Last modification timestamp of the region", example = "2023-01-01T12:00:00")
+    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd'T'HH:mm:ss")
+    @JsonProperty("modified_on")
+    private LocalDateTime modifiedOn;
+
+    @Schema(description = "User who last modified the region", example = "admin")
+    @JsonProperty("modified_by")
+    private String modifiedBy;
+
```
Comment (NdubuisiJr) [inline src/main/java/hng_java_boilerplate/region/dto/RegionDto.java:49]: Remove extra space

File: src/main/java/hng_java_boilerplate/region/dto/UserRegionDto.java [hunk 13]
```diff
@@ -0,0 +1,36 @@
+package hng_java_boilerplate.region.dto;
+
+
```
Comment (NdubuisiJr) [inline src/main/java/hng_java_boilerplate/region/dto/UserRegionDto.java:3]: Remove extra space

File: src/main/java/hng_java_boilerplate/region/entity/RegionEntity.java [hunk 14]
```diff
@@ -0,0 +1,56 @@
+package hng_java_boilerplate.region.entity;
+
```
Comment (NdubuisiJr) [inline src/main/java/hng_java_boilerplate/region/entity/RegionEntity.java:2]: Remove extra space

File: src/main/java/hng_java_boilerplate/region/entity/UserRegionEntity.java [hunk 15]
```diff
@@ -0,0 +1,38 @@
+package hng_java_boilerplate.region.entity;
+
```
Comment (NdubuisiJr) [inline src/main/java/hng_java_boilerplate/region/entity/UserRegionEntity.java:2]: Remove extra space

File: src/main/java/hng_java_boilerplate/region/repository/RegionRepository.java [hunk 16]
```diff
@@ -0,0 +1,12 @@
+package hng_java_boilerplate.region.repository;
+
```
Comment (NdubuisiJr) [inline src/main/java/hng_java_boilerplate/region/repository/RegionRepository.java:2]: Remove extra spaces

File: src/main/java/hng_java_boilerplate/region/repository/UserRegionRepository.java [hunk 17]
```diff
@@ -0,0 +1,16 @@
+package hng_java_boilerplate.region.repository;
+
+
```
Comment (NdubuisiJr) [inline src/main/java/hng_java_boilerplate/region/repository/UserRegionRepository.java:3]: Remove extra space

File: src/main/java/hng_java_boilerplate/region/service/RegionService.java [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+        }
+    }
+
+    public boolean hasUserAssignedRegion(UUID userId) {
+        Optional<UserRegionEntity> userRegion = userRegionRepository.findByUserId(userId);
+        return userRegion.isPresent();
+    }
+}
+
+// this comment is to check if ci cd test can be triggered.
```
Comment (NdubuisiJr) [inline src/main/java/hng_java_boilerplate/region/service/RegionService.java:124]: Remove this comment

**Review by NdubuisiJr — CHANGES_REQUESTED** [review]

**Comment by unyimeudemy** [comment]
@NdubuisiJr All unnecessary spaces have been removed

**Review by NdubuisiJr — APPROVED** [review]

**Pull request closed without merging**